Plugins/NativeXr/Source/NativeXr.cpp

```diff
@@ -301,296 +301,342 @@ namespace
+                return itFrameBufferToJsTexture->second.Value();
+            }
 
-        void SetPlaneDetectionEnabled(bool enabled)
-        {
-            m_session->SetPlaneDetectionEnabled(enabled);
-        }
+            void SetDepthsNarFar(float depthNear, float depthFar)
+            {
+                m_sessionState->session->SetDepthsNearFar(depthNear, depthFar);
+            }
 
-        bool TrySetFeaturePointCloudEnabled(bool enabled)
-        {
-            return m_session->TrySetFeaturePointCloudEnabled(enabled);
-        }
+            void SetPlaneDetectionEnabled(bool enabled)
+            {
+                m_sessionState->session->SetPlaneDetectionEnabled(enabled);
+            }
 
-        bool TrySetPreferredPlaneDetectorOptions(const xr::GeometryDetectorOptions& options)
-        {
-            return m_session->TrySetPreferredPlaneDetectorOptions(options);
-        }
+            bool TrySetFeaturePointCloudEnabled(bool enabled)
+            {
+                return m_sessionState->session->TrySetFeaturePointCloudEnabled(enabled);
+            }
+
+            bool TrySetPreferredPlaneDetectorOptions(const xr::GeometryDetectorOptions& options)
+            {
+                return m_sessionState->session->TrySetPreferredPlaneDetectorOptions(options);
+            }
+
+            bool TrySetMeshDetectorEnabled(const bool enabled)
+            {
+                return m_sessionState->session->TrySetMeshDetectorEnabled(enabled);
+            }
+
+            bool TrySetPreferredMeshDetectorOptions(const xr::GeometryDetectorOptions& options)
+            {
+                return m_sessionState->session->TrySetPreferredMeshDetectorOptions(options);
+            }
+
+            uintptr_t GetNativeXrContext()
+            {
+                return m_sessionState->session->GetNativeXrContext();
+            }
+
+            std::string GetNativeXrContextType()
+            {
+                return m_sessionState->session->GetNativeXrContextType();
+            }
+
+        private:
+            Napi::Env m_env;
+            JsRuntimeScheduler m_runtimeScheduler;
+            std::function<void(bool)> m_sessionStateChangedCallback{};
+            void* m_windowPtr{};
+
+            struct SessionState
+            {
+                explicit SessionState(Graphics::Impl& graphicsImpl)
+                    : graphicsImpl{ graphicsImpl }
+                {
+                }
+
+                Graphics::Impl& graphicsImpl;
+                Napi::FunctionReference createRenderTexture{};
+                Napi::FunctionReference destroyRenderTexture{};
+                std::map<void*, FrameBuffer*> textureToFrameBufferMap{};
+                std::map<FrameBuffer*, Napi::ObjectReference> frameBufferToJsTextureMap{};
+                std::vector<void*> activeTextures{};
+                xr::System system{};
+                std::shared_ptr<xr::System::Session> session{};
```

Comment:
should the system be owned by the session state or should it also exist for the duration of the js engine like the xr napi component?

Reply:
this comment/ask shouldn't be considered blocking. it does seem like now if we've wanted to expose xr system specific functionality with no tie to a session (like enabling OpenXR extensions prior to the xr session creation), that the system could now be owned by NativeXr::Impl

---
In reply to: [593313639](https://github.com/BabylonJS/BabylonNative/pull/631#discussion_r593313639) [](ancestors = 593313639)

Reply:
talked offline, sounds like we'll plan for this in a follow up review

Reply:
I think it should not be part of the session state, and should live on the `Napi::Impl` directly, but this needs to be tested on HoloLens since `XR::System` actually does something on that platform (on iOS and Android it basically does nothing). Discussed more offline, and @chrisfromwork will test this out for a separate PR maybe next week.